logd/server/collector.go

```diff
@@ -258,35 +265,110 @@ func (lc *LogCollector) getEvents(ctx context.Context, input *api.GetEventsInput
 		}
 	}
 
-	// TODO: Allow override via input.
-	limit := defaultEventsPerRequest
+	limit := defaultLimit
+	var direction store.Direction
+	if input.Next != nil {
+		if input.Prev != nil {
+			return nil, errors.New("Only one of prev or next may be specified")
+		}
+		limit = *input.Next
+		direction = store.DirectionForward
+	} else if input.Prev != nil {
+		limit = *input.Prev
+		direction = store.DirectionBackward
+	} else {
+		// Use default limit, and move forward
+		direction = store.DirectionForward
+	}
+
+	var cursor *store.Cursor
+	var err error
+	if input.Cursor == nil {
+		if cursor, err = lc.getLatestCursor(ctx, input.Logs); err != nil {
+			return nil, fmt.Errorf("finding latest cursor: %w", err)
+		}
+	} else {
+		if cursor, err = store.ParseCursor(*input.Cursor); err != nil {
+			return nil, fmt.Errorf("parsing cursor: %w", err)
+		}
+	}
 
 	// TODO: Merge sort.
 	events := []api.Event{}
 	for _, logName := range logs {
 		log := lc.store.GetLog(logName)
-		logEvents, err := log.GetEvents(ctx, input.After, limit)
+
+		logCursor := cursor
+		if logCursor == nil {
+			if lastEvent := log.GetLastEvent(ctx); lastEvent != nil {
+				idBytes, err := badger.DecodeID(lastEvent.ID)
+				if err != nil {
+					return nil, fmt.Errorf("decoding id: %w", err)
+				}
+				idBytes = binaryutil.IncrementBytes(idBytes)
+				logCursor = &store.Cursor{ID: idBytes}
+			}
+		}
+
+		logEvents, err := log.GetEvents(ctx, logCursor, limit, direction)
 		if err != nil {
 			return nil, fmt.Errorf("getting %q events: %w", logName, err)
 		}
+		if logEvents == nil {
+			continue
+		}
 		events = append(events, logEvents...)
 	}
 	sort.Sort(&eventsSorter{events})
 
-	var cursor string
+	effectiveLimit := mathutil.IntMin(limit, len(events))
+	if direction == store.DirectionForward {
+		events = events[0:effectiveLimit]
+	} else {
+		end := len(events)
+		start := end - effectiveLimit
+		events = events[start:end]
+	}
+
+	prevCursor := cursor
+	nextCursor := cursor
 	if len(events) > 0 {
-		end := mathutil.IntMin(limit, len(events))
-		events = events[:end]
+		// TODO: Separate encoding from badger code.
+		if idBytes, err := badger.DecodeID(events[0].ID); err != nil {
+			return nil, err
+		} else {
+			// We don't care about the error if this is already a (zero-valued) NilCursor
+			_ = binaryutil.DecrementBytes(idBytes)
```

Comment:
push the IncrementBytes and DecrementBytes calls down in to the Log struct & change GetEvents to return cursors.

the log collector level should not be mucking with the cursors' internals.